feat: backend to send revenue claim tx

troggy opened the pull request:
## What does this PR do?

- create extra HTTP endpoint to queue "revenue claim" tx against Collection Pool contract
- uses the same priv key as XDAI faucet
- share sender lambda for faucet and revenue claim
- revenue claim sending lambda doesn't await for tx to be mined and subscribes for tx confirmation via Blocknative Notify.js API
- add pusher lambda as a webhook for Blocknative Notify.js. When tx landed (or failed) pushes event to Pusher to notify client (see https://github.com/WhiteRabbitTheWorldIsYours/dashboard/pull/74)

## Related issues
Requires: https://github.com/WhiteRabbitTheWorldIsYours/dashboard/pull/103
Relates: https://github.com/WhiteRabbitTheWorldIsYours/dashboard/pull/74
Closes #4 
Closes #5

johannbarbie commented on src/revenue/sender.js:
```diff
@@ -0,0 +1,39 @@
+  return null;
+};
+
+exports.handler = async (event) => {
```
doesn't the payout need to be marked as done in some database? otherwise things can be claimed twice?

johannbarbie commented on src/revenue/registerRequest.js:
```diff
@@ -0,0 +1,126 @@
+console.log(queue);
+const imdbToTokenId = require('./utils/imdbToTokenId');
+
+const THE_GRAPH_URL = 'https://api.thegraph.com/subgraphs/name/whiterabbittheworldisyours/collectionpool';
```
move to serverless.yml as config param?

johannbarbie commented on src/revenue/registerRequest.js:
```diff
@@ -0,0 +1,126 @@
+const imdbToTokenId = require('./utils/imdbToTokenId');
+
+const THE_GRAPH_URL = 'https://api.thegraph.com/subgraphs/name/whiterabbittheworldisyours/collectionpool';
+const BE_URL = 'https://be-dashboard.whiterabbit.one/graphql';
```
move to serverless.yml as config param?

johannbarbie commented on src/revenue/registerRequest.js:
```diff
@@ -0,0 +1,126 @@
+  await checkAuthorized(movieId, signer);
+
+  const movieTokenId = imdbToTokenId(movieId);
+  const revenue = await getUnclaimedRevenue(movieTokenId, region);
```
the graph is updated very sluggishly: tx pending for x seconds + scanner only running every x seconds.
what stops a rightsholder from calling the function many times at once, and getting things paid out multiple times, before the graph actually returns 0?

johannbarbie commented on src/revenue/registerRequest.test.js:
```diff
@@ -0,0 +1,158 @@
+  );
+});
+
+it('should enqueue a claim request', async () => {
```
❤️

johannbarbie commented on src/revenue/registerRequest.test.js:
```diff
@@ -0,0 +1,158 @@
+});
+
+
+it('should fail if there is 0 unclaimed revenue for a given movie', async () => {
```
❤️

johannbarbie commented on src/revenue/sender.js:
```diff
@@ -0,0 +1,52 @@
+    })
+      .then((tx) => {
+        console.log('Distribution', request.key, tx.hash);
+        return db.setDistributionTime(request.key);
```
isn't this missing an `await` ?

troggy replied:
no, it is a `Promise` returned from `then` and we will await for it on the line 43

johannbarbie commented on src/revenue/utils/db.js:
```diff
@@ -0,0 +1,18 @@
+const { SimpleDb } = require('leap-lambda-boilerplate');
```
the boilerplate doesn't allow to inject dependencies, hence unit-testing will not be possible on external interfaces, rather will require to mock internal interfaces (antipattern). why not add the functions to existing db? https://github.com/WhiteRabbitTheWorldIsYours/payment-backend/blob/master/src/utils/db.js

troggy replied:
> hence unit-testing will not be possible on external interfaces

it will be possible. You can mock `setAttr` and `getAttr` methods of the `SimpleDb` (which is an external interface here)

troggy replied:
> why not add the functions to existing db? 

just so we don't mix concerns. I've already mixing it enough by having two unrelated things (faucet and claims) together in a single code base. Besides, I don't like boilerplate in existing db and would rather rewrite it to use `SimpleDb` abstraction as well :)

johannbarbie replied:
> You can mock setAttr and getAttr methods of the SimpleDb

cool, i wasn't aware that is possible. looking forward to learn it.

troggy replied:
in the end I managed to make db-less tests without mocking database calls (DB.js remains untested, but it is very small). Still possible to test it though, let me know if you are curious to know how — I can give you a demo.

On a second note, we should stop using sinon in favour of jest. It is much easier to do mocks with Jest

johannbarbie commented:
i have the following errors on running tests:

```
      -Bad Request: Missing movieId or regionId in the URL
      +Unauthorized: Missing authentication token
    
      -Bad Request: Missing movieId or regionId in the URL
      +Unauthorized: Authentication token has expired
    
      -Bad Request: Missing movieId or regionId in the URL
      +Bad Request: No such movie
    
      -Bad Request: Missing movieId or regionId in the URL
      +Forbidden: Not a rightsholder
    
      -Bad Request: Missing movieId or regionId in the URL
      +Bad Request: Nothing to distribute
```

johannbarbie commented on src/revenue/registerRequest.js:
```diff
@@ -0,0 +1,126 @@
+  if (revenue === BigInt(0)) {
+    throw new Errors.BadRequest('Nothing to distribute');
+  }
+
```
it would be advantage to also check the db lock here, so that we can respond with an http error code if the payout is called to early. would simplify frontend debugging.

troggy replied:
✅

troggy commented:
tests not working great together with the old one. Figuring out

johannbarbie commented on src/revenue/registerRequest.test.js:
```diff
@@ -0,0 +1,188 @@
+    );
+  });
+  
+  it('should fail if there if distribution was requested too soon after the previous one', async () => {
```
❤️

troggy commented:
deployed on AWS with min frequency of 30 seconds

troggy commented:
collection pool is still owned by other account, so all the claims will fail. I've verified sender by commenting out sending code

johannbarbie commented:
> collection pool is still owned by other account, so all the claims will fail. I've verified sender by commenting out sending code

... checking.

johannbarbie commented:
added faucet address as owner in pool: https://blockscout.com/poa/xdai/tx/0x72b089073185875e20d327ba0560d20987a990e9269f9e25cc6712a20e02b940/internal_transactions

troggy commented on src/revenue/sender.js:
```diff
@@ -33,23 +33,28 @@ exports.handler = async (event) => {
-  for (let request of requests) {
-    const { result, minutesTillNextDistribution } = await canDistributeNow(request.key, minFrequency, db);
+  let nonce = (await wallet.getTransactionCount()) - 1;
+  await Promise.all(requests.map(async (request) => {
```
can do this because I'm managing nonce now. It is faster — multiple txs could be submitted simultatenously and possible make it to the same block, while with `for .. of` loop you have to wait for each of the txs to be mined (N blocks)

troggy commented:
@johannbarbie pushed couple of things to make it work with the frontend + race condition fixes. Let me know when it is good for merge

troggy commented:
@johannbarbie can I merge this?

johannbarbie commented:
checking

johannbarbie commented on src/revenue/sender.js:
```diff
@@ -33,23 +33,28 @@ exports.handler = async (event) => {
-        nonce = nonce + 1;
-        return db.setDistributionTime(request.key);
+        return db.setDistribution(request.key, request.uuid, tx.hash)
+          .then(() => tx.wait());
```
wouldn't this exceed lambda runtime duration?

troggy replied:
depending on the network block time, yes it could. You certainly don't want to run this on ethereum mainnet. I can rewrite it to store `nonce` in database, so we don't need to wait for them to be mined. However, this complicates pusher task — we will need to somehow check if older txs are executed so you can make a push to the client

johannbarbie replied:
i guess i don't understand how the await is affected by the nonce.
- `requests.map()` initializes the nonces in sequence, but
- `Promise.all()` waits for all transaction submission in parallel. 
so let's say that submitting tx 2 out of 3 would fail, then 3 would be published, but never land, as it's nonce is not in sequence, right?
how does `tx.wait()` chainge that?

johannbarbie replied:
i guess it is important to put those requests that can not be fulfilled back into SQS, then they can be retried. if we would only take 1 element from SQS, it would be even better, then the promise-all would not be needed.

troggy replied:
> so let's say that submitting tx 2 out of 3 would fail, then 3 would be published, but never land, as it's nonce is not in sequence, right?

🤔 this is a good point

johannbarbie replied:
on the other hand, not sure how a transaction submission can fail, other then network problems. if it is retried, then the setup should be sound.

troggy replied:
> wouldn't this exceed lambda runtime duration?

solved by not waiting for confirmation anymore. To subscribe for confirmation I delegate to Blocknative Notify.js

troggy commented:
Reworked sender: now using the same lambda both for faucet and revenue claim (as they share a nonce) + subscribe for tx confirmation via Blocknative API.

Added an extra webhook lambda which is triggered by Blocknative. This lambda pushes update to the client. 

[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVEJcblx0QVtjbGllbnQgZnJvbnRlbmRdOjo6Y2xpZW50IC0tPiBCKGNsYWltIHJlcXVlc3QgcmVnaXN0ZXIgbGFtYmRhKVxuXHRDWyhEQildIC0uLSBCXG4gICAgRFsoVGhlR3JhcGgpXTo6OmV4dGVybmFsIC0uLT4gQlxuXHRCIC0tPiBFKFwic2VuZGVyIGxhbWJkYSAoc2luZ2xldG9uKVwiKVxuXHRDIC0uLSBFXG5cbiAgICBFIC0tPiBGWyh4ZGFpIGNoYWluKV06OjpleHRlcm5hbFxuICAgIEUgLS0-IEcoW0Jsb2NrbmF0aXZlLmNvbV0pOjo6ZXh0ZXJuYWxcbiAgICBHIC0tPiBIW3B1c2hlciBsYW1iZGFdXG4gICAgQyAtLi0gSFxuICAgIEggLS0-IEkoW3B1c2hlci5jb21dKTo6OmV4dGVybmFsXG4gICAgSSAtLT4gQVxuICAgIGNsYXNzRGVmIGV4dGVybmFsIGZpbGw6I2ZmZWVkZDtcbiAgICBjbGFzc0RlZiBjbGllbnQgZmlsbDojZWVmZmRkO1xuXG5cblx0XHRcdFx0XHQiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVEJcblx0QVtjbGllbnQgZnJvbnRlbmRdOjo6Y2xpZW50IC0tPiBCKGNsYWltIHJlcXVlc3QgcmVnaXN0ZXIgbGFtYmRhKVxuXHRDWyhEQildIC0uLSBCXG4gICAgRFsoVGhlR3JhcGgpXTo6OmV4dGVybmFsIC0uLT4gQlxuXHRCIC0tPiBFKFwic2VuZGVyIGxhbWJkYSAoc2luZ2xldG9uKVwiKVxuXHRDIC0uLSBFXG5cbiAgICBFIC0tPiBGWyh4ZGFpIGNoYWluKV06OjpleHRlcm5hbFxuICAgIEUgLS0-IEcoW0Jsb2NrbmF0aXZlLmNvbV0pOjo6ZXh0ZXJuYWxcbiAgICBHIC0tPiBIW3B1c2hlciBsYW1iZGFdXG4gICAgQyAtLi0gSFxuICAgIEggLS0-IEkoW3B1c2hlci5jb21dKTo6OmV4dGVybmFsXG4gICAgSSAtLT4gQVxuICAgIGNsYXNzRGVmIGV4dGVybmFsIGZpbGw6I2ZmZWVkZDtcbiAgICBjbGFzc0RlZiBjbGllbnQgZmlsbDojZWVmZmRkO1xuXG5cblx0XHRcdFx0XHQiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)

troggy commented:
This is working with the UI from here https://github.com/WhiteRabbitTheWorldIsYours/dashboard/pull/74

troggy commented:
diagram is generated from `claimArchitecture.mmd` file 🤓

troggy commented on package.json:
```diff
@@ -4,8 +4,8 @@
-    "lint": "eslint --fix src/*.js src/utils/*.js",
-    "test": "mocha src/*.test.js src/**/*.test.js",
+    "lint": "eslint --fix src/**/*.js",
+    "test": "./node_modules/.bin/mocha --recursive 'src/**/*.test.js'",
```
this properly recurses into subfolders of any depth

johannbarbie replied:
thx, it was driving me crazy 🤣

troggy commented on package.json:
```diff
@@ -4,8 +4,8 @@
-    "lint": "eslint --fix src/*.js src/utils/*.js",
-    "test": "mocha src/*.test.js src/**/*.test.js",
+    "lint": "eslint --fix src/**/*.js",
+    "test": "./node_modules/.bin/mocha --recursive 'src/**/*.test.js'",
```
```suggestion
    "test": "mocha --recursive 'src/**/*.test.js'",
```

troggy commented on src/faucet/sender.js:
```diff
@@ -0,0 +1,31 @@
+/**
```
this file was just moved into `faucet` folder. No changes

troggy commented on src/revenue/pusher.js:
```diff
@@ -0,0 +1,54 @@
+      result: resultMapping[status] || 0,
+      txHash: hash,
+    });
+    await db.deleteWatchedTx(hash, regionId, movieId);
```
this seems not working (records stay in database). Not critical, but I'll look into this separately (already blew the scope)

johannbarbie commented on README.md:
```diff
@@ -179,20 +179,24 @@ Response 201:
 
+## Claim Revenue Architecture
+
+[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVEJcblx0QVtjbGllbnQgZnJvbnRlbmRdOjo6Y2xpZW50IC0tPiBCKGNsYWltIHJlcXVlc3QgcmVnaXN0ZXIgbGFtYmRhKVxuXHRDWyhEQildIC0uLSBCXG4gICAgRFsoVGhlR3JhcGgpXTo6OmV4dGVybmFsIC0uLT4gQlxuXHRCIC0tPiBFKFwic2VuZGVyIGxhbWJkYSAoc2luZ2xldG9uKVwiKVxuXHRDIC0uLSBFXG5cbiAgICBFIC0tPiBGWyh4ZGFpIGNoYWluKV06OjpleHRlcm5hbFxuICAgIEUgLS0-IEcoW0Jsb2NrbmF0aXZlLmNvbV0pOjo6ZXh0ZXJuYWxcbi ...
```
wow, never thought about linking an image in markdown 🤔 
...and then directly link to the same thing in the editor, cool 👍

johannbarbie replied:
[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVEJcblx0QVtyaWdodHNob2xkZXIgZnJvbnRlbmRdOjo6Y2xpZW50XG4gICAgTFt2aWV3ZXIgZnJvbnRlbmRdOjo6Y2xpZW50XG5cbiAgICBEWyhUaGVHcmFwaCldOjo6ZXh0ZXJuYWxcbiAgICBGWyh4ZGFpIGNoYWluKV06OjpleHRlcm5hbFxuICAgIEcoW0Jsb2NrbmF0aXZlLmNvbV0pOjo6ZXh0ZXJuYWxcbiAgICBJKFtwdXNoZXIuY29tXSk6OjpleHRlcm5hbFxuICAgIFAoW2NpcmNsZS5jb21dKTo6OmV4dGVybmFsXG5cbiAgICBCKFwicmV2ZW51ZUNsYWltKClcIilcbiAgICBLKFwiZmF1Y2V0Q2xhaW0oKVwiKVxuICAgIEhbXCJwdXNoZXIoKVwiXVxuICAgIEUoXCJzZW5kZXIoKVwiKTo6OnNpbmdsZXRvblxuICAgIE8oXCJjaXJjbGVQcm94eSgpXCIpXG5cbiAgICBKKFwic2VuZGVyIFNRU1wiKVxuICAgIE4oXCJmYXVjZXQgU1FTXCIpXG4gICAgXG5cbiAgICBDWyhSZXZlbnVlIERCKV1cbiAgICBNWyhGYXVjZXQgREIpXVxuXG5cbiAgICBBIC0tPiBCXG5cdEMgLS4tIEJcbiAgICBEIC0uLT4gQlxuXHRDIC0uLSBFXG4gICAgRSAtLi0gTVxuXG4gICAgRSAtLT4gRlxuICAgIEUgLS0-IEdcbiAgICBHIC0tPiBIXG4gICAgQyAtLi0gSFxuICAgIEggLS0-IElcbiAgICBJIC0uLT4gQVxuICAgIEIgLS0-IEpcbiAgICBKIC0tPiBFXG5cbiAgICBMIC0tPiBLXG4gICAgSyAtLT4gTlxuICAgIE4gLS0-IEVcbiAgICBJIC0uLT4gTFxuICAgIEwgLS0-IE9cbiAgICBPLS1jcmVhdGVQYXltZW50LS0-UFxuICAgIE8gLS4tPiBNXG4gICAgTSAtLi0-IEtcblxuICAgIGNsYXNzRGVmIGV4dGVybmFsIGZpbGw6I2ZmZWVkZDtcbiAgICBjbGFzc0RlZiBzaW5nbGV0b24gZmlsbDojZmZkZGVlO1xuICAgIGNsYXNzRGVmIGNsaWVudCBmaWxsOiNlZWZmZGQ7XG5cblxuXHRcdFx0XHRcdCIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVEJcblx0QVtyaWdodHNob2xkZXIgZnJvbnRlbmRdOjo6Y2xpZW50XG4gICAgTFt2aWV3ZXIgZnJvbnRlbmRdOjo6Y2xpZW50XG5cbiAgICBEWyhUaGVHcmFwaCldOjo6ZXh0ZXJuYWxcbiAgICBGWyh4ZGFpIGNoYWluKV06OjpleHRlcm5hbFxuICAgIEcoW0Jsb2NrbmF0aXZlLmNvbV0pOjo6ZXh0ZXJuYWxcbiAgICBJKFtwdXNoZXIuY29tXSk6OjpleHRlcm5hbFxuICAgIFAoW2NpcmNsZS5jb21dKTo6OmV4dGVybmFsXG5cbiAgICBCKFwicmV2ZW51ZUNsYWltKClcIilcbiAgICBLKFwiZmF1Y2V0Q2xhaW0oKVwiKVxuICAgIEhbXCJwdXNoZXIoKVwiXVxuICAgIEUoXCJzZW5kZXIoKVwiKTo6OnNpbmdsZXRvblxuICAgIE8oXCJjaXJjbGVQcm94eSgpXCIpXG5cbiAgICBKKFwic2VuZGVyIFNRU1wiKVxuICAgIE4oXCJmYXVjZXQgU1FTXCIpXG4gICAgXG5cbiAgICBDWyhSZXZlbnVlIERCKV1cbiAgICBNWyhGYXVjZXQgREIpXVxuXG5cbiAgICBBIC0tPiBCXG5cdEMgLS4tIEJcbiAgICBEIC0uLT4gQlxuXHRDIC0uLSBFXG4gICAgRSAtLi0gTVxuXG4gICAgRSAtLT4gRlxuICAgIEUgLS0-IEdcbiAgICBHIC0tPiBIXG4gICAgQyAtLi0gSFxuICAgIEggLS0-IElcbiAgICBJIC0uLT4gQVxuICAgIEIgLS0-IEpcbiAgICBKIC0tPiBFXG5cbiAgICBMIC0tPiBLXG4gICAgSyAtLT4gTlxuICAgIE4gLS0-IEVcbiAgICBJIC0uLT4gTFxuICAgIEwgLS0-IE9cbiAgICBPLS1jcmVhdGVQYXltZW50LS0-UFxuICAgIE8gLS4tPiBNXG4gICAgTSAtLi0-IEtcblxuICAgIGNsYXNzRGVmIGV4dGVybmFsIGZpbGw6I2ZmZWVkZDtcbiAgICBjbGFzc0RlZiBzaW5nbGV0b24gZmlsbDojZmZkZGVlO1xuICAgIGNsYXNzRGVmIGNsaWVudCBmaWxsOiNlZWZmZGQ7XG5cblxuXHRcdFx0XHRcdCIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)

johannbarbie commented on README.md:
````diff
@@ -202,19 +206,28 @@ Create required properties in [AWS parameter store](https://eu-west-1.console.aw
-```
+
+2. (First deployment only) Set up `dispenseTokens` lambda to run on new SQS message.
+⚠️ It is a unsolved issue in serverless config — it should be set up with serverless, but isn't working for some reason.
````
🤔

troggy replied:
probably not an issue anymore, since you were able to deploy it the very first time without this manual step

troggy commented:
still reviewing? Sorry, got confused by your approve. Should I reopen it?

johannbarbie commented:
i guess i miss-clicked. still digging through it. i'll re-open if i find anything.
